javascript/react-js/react-testing-part-two.md

```diff
@@ -0,0 +1,160 @@
+
+#### submissions-list.test.jsx
+
+We notice there are two child components of `SubmissionsList`. One of them is from a package called `react-flip-move`. External Code. We'll mock it. 
+
+<span id="mock-child-component">Notice how we mock the `Submission` component</span>:
+
+~~~javascript 
+jest.mock('../submission', () => ({ submission, isDashboardView }) => (
+  <>
+    <div data-test-id="submission">{submission.id}</div>
+    <div data-test-id="dashboard">{isDashboardView.toString()}</div>
+  </>
+));
+~~~
+
+We only render the bare minimum to realize the validity of the component we're testing. Next, we set up our props with fake data and mocked functions.
+
+Let's move towards our first assertion. Don't worry too much about the `ProjectSubmissionContext.Provider`. In the context of this test, it's purpose is to act as a route to pass in the `allSubmissionsPath` prop. We've already identified the three points of interest that we want to test. We divide them into three test suites for readability purposes using `describe`.
+
+In the first suite, we make some assertions if the user has a submission and then some assertions if the user does not. The other suites follow a similar pattern.
+
+#### Exploring Further
+
+Feel free to flick through the other components and its tests. You'll see mocked functions. You'll see some unseen functions like `act`. You'll see custom render functions. It's fine if you don't understand it all, the goal is to gain familiarity.
+
+The other important thing to note is almost all the tests follow a certain pattern in terms of the way they're written. They follow the [Arrange-Act-Assert](http://wiki.c2.com/?ArrangeActAssert) pattern. It's a good idea to adopt the pattern sooner or later to make your tests more readable and ultimately better.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Read about the [secrets of act(...) api](https://github.com/mrdulin/react-act-examples/blob/master/sync.md). Especially pay attention to the last bit about testing asynchronous code. You won't be using it everyday, but it's nice to be familiar.
+
+2. Read this article about [mocking child components](https://medium.com/@taylormclean15/jest-testing-mocking-child-components-to-make-your-unit-tests-more-concise-18691ef6a0c2). It extensively covers the "how" of mocking child components. (You might need to sign up to read)
+ 
+</div>
+
+### Knowledge Check
+
+This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.
+
+* <a class="knowledge-check-link" href="#testing-callback-handlers">How can you mock a callback handler?</a>
+* <a class="knowledge-check-link" href="#mock-child-component">How can you mock a child component?</a>
```

Comment:
I could just not be seeing it, but it doesn't look like this jump link corresponds with any currently existing section?

Reply:
The "How can you mock a child component?" knowledge check should jump to line 116 which is:

```html
<span id="mock-child-component">Notice how we mock the `Submission` component</span>:
```

Clicking the link should jump to this line, I've tested it with the Lesson Preview Tool

Reply:
👍 yep I just missed it.